### ISSUE TITLE
Compiler hangs for plausible, yet very specific scenario

### ISSUE THREAD

**Issue by ricardopieper** [luna/luna#116]
TL;DR; The texts in **bold** indicate where I found a case where Luna just gets stuck while compiling.

I was writing a bit of code for a string split function. At some moment Luna just couldn't compile my code anymore.

Specifically, it just gets stuck here:

`Compiling module: Split.Main`

With one of my CPU cores locked at 100%.

So I retraced my steps and came up with a minimal example that reproduces it:

```
def main:
    finished = typeInferenceBug "a"
    print finished
    None
    
def typeInferenceBug a:
    if a == "a" then [a] else [[a]]
```

First I named the function `typeInferenceBug` because I thought that's what was going on. The function clearly doesn't pass the type checker due to to the conflicting returns ([[T]] and [T]).

It tried to compile this little code for like 5 minutes. That's more than enough time to compile 6 lines of code, so its probably stuck in a loop.

Exploring the bug a little bit more, this does not reproduce the bug, and Luna behaves correctly:

```
def main:
    finished = doesNotCompile "a"
    print finished
    None
    
def doesNotCompile a:
    if a == "a" then 1 else "a"
```

Unification error: (Text) with (Int)
    Arising from:
        ↳  Split.Main : doesNotCompile
        ↳  Split.Main : main

Maybe it's just for algebraic types? Work with me through the next samples. 

I'll make a new Maybe class, just naming it differently to not clash with the existing one:

```
def main:
    finished = typeInferenceBug "a"
    print "finished"
    None

def typeInferenceBug a:
    if a == "a" then Present (Present "a") else Present 1

class Nullable a:
    Present a
    Null
```

This works correctly:
```
 Unification error: (Int) with (Nullable(Text))
    Arising from:
        ↳  Split.Main : typeInferenceBug
        ↳  Split.Main : main
```

Changing Present "a" to Present 1 doesn't reproduce the bug either. However, List is a little bit more complex, lets try using List without the List [] syntax:

```
def main:
    finished = typeInferenceBug "a"
    print "finished"
    None

def typeInferenceBug a:
    if a == "a" then Prepend (Prepend 1 Empty) Empty else Prepend 1 Empty
```

Nope, luna behaves correctly, this works just fine:

```
• Unification error: (Int) with (List(Int))
    Arising from:
        ↳  Split.Main : typeInferenceBug
        ↳  Split.Main : main
```

Let's use the list syntax:

```
def main:
    finished = listSyntaxBug "a"
    print "finished"
    None

def listSyntaxBug a:
    if a == "a" then [[1]] else [1]
```
That also works. But it's exactly the same as the first code snippet! So what gives?

Well, it's not exactly the same. In the first code snippet I'm returning the parameter "a" inside the list, so I'll use the "a" parameter now, just like before:

```
def main:
    finished = weirdBug "a"
    print "finished"
    None

def weirdBug a:
    if a == "a" then [[a]] else [a]
```
**This reproduces the bug**. So, **returning the list with the parameter** breaks the compiler. Let's go back to what I did before with the Nullable class:

```
def main:
    finished = typeInferenceBug "a"
    print "finished"
    None

def typeInferenceBug a:
    if a == "a" then Present (Present a) else Present a

class Nullable a:
    Present a
    Null
```

**This also reproduces the bug.**

And so does this, using Prepends:

```
def typeInferenceBug a:
    if a == "a" then Prepend (Prepend a Empty) Empty else Prepend a Empty
```

Let's change just one of them to a constant literal value to see what happens:

```
def typeInferenceBug a:
    if a == "a" then Prepend (Prepend a Empty) Empty else Prepend "a" Empty
```
It works correctly:

```
• Unification error: (Text) with (List(#b))
    Arising from:
        ↳  Split.Main : typeInferenceBug
        ↳  Split.Main : main
```

Changing from "a" to 1 yields the same result.

Let's try some more:

```
def typeInferenceBug a:
    b = 1
    if a == "a" then Prepend (Prepend b Empty) Empty else Prepend b Empty
```

This doesn't reproduce the bug, it works fine. But `b` is a local literal value, and I guess its type is easily accessible. How about this one?

```
def returnsSomething a: 
    1

def typeInferenceBug a:
    b = returnsSomething a
    if a == "a" then Prepend (Prepend b Empty) Empty else Prepend b Empty 
```

**That reproduces the bug as well**. I think Luna thinks the type of `b` probably depends on the type of `a`, maybe? But clearly not: it just returns a literal number, whose type is/should be well known. I know nothing about type systems though, maybe I'm oversimplifying it.

Anyway, that's all the info I have! Notice that this is a thing that happened on Luna Studio when I was writing a more real function. It's not a crazy scenario when you're just experimenting with new code.

**Comment by ricardopieper** [comment]
I cloned the luna repo and i'm running the compiler in the command line btw. But I had the same thing happening in luna editor.

**Comment by iamrecursion** [comment]
I can reproduce this.

**Comment by kustosz** [comment]
Thank you for the report @ricardopieper, great analysis! The compiler loses at the infinite-type game here. This needs a fix rather quickly.
The `a` in the original repro is polymorphic and the return value generates an equation of the form:
```
[a] = [[a]]
```
Which then simplifies to `a = [a]`, and then gets "solved" by substitution, so it evolves to `a = [[a]]`, then to `a = [[[a]]]` and so on. This is a rather quick fix, for someone who'd like to get a gist of how the compiler works. We need to detect this particular shape and throw a "cannot construct infinite type" error.

**Comment by piotrMocz** [comment]
@kustosz is that fixed?
